### PR TITLE
Include bundled aggregations in /sync and related fixes

### LINKED ISSUE #11476: /sync is not bundling aggregations for limited rooms
[MSC2675](https://github.com/matrix-org/matrix-doc/pull/2675) says that `/sync` is one of the APIs which should bundle aggregations:

> `GET /sync`, only for room sections in the response where `limited` field is `true`; this amounts to all rooms in the response if the `since` request parameter was not passed, also known as an initial sync.

Currently it always disables bundled aggregations for this endpoint, see #5220.

We should also double check the other endpoints:

* [x] `GET /rooms/{roomId}/messages` -- [see code](https://github.com/matrix-org/synapse/blob/a265fbd397ae72b2d3ea4c9310591ff1d0f3e05c/synapse/handlers/pagination.py#L545-L558)
* [x] `GET /rooms/{roomId}/context` -- [see code](https://github.com/matrix-org/synapse/blob/a265fbd397ae72b2d3ea4c9310591ff1d0f3e05c/synapse/rest/client/room.py#L709-L723) (note that this doesn't bundle them for state events, unclear if this is "fine" or not)
* [x] `GET /rooms/{roomId}/event/{eventId}` -- [see code](https://github.com/matrix-org/synapse/blob/a265fbd397ae72b2d3ea4c9310591ff1d0f3e05c/synapse/rest/client/room.py#L664-L666)
* [ ] `GET /sync`
* [x] `GET /relations` -- see #11087

### PR DESCRIPTION
**Pull request by clokep**
This has a few fixes related to *when* we include bundled aggregations:

* Renames `bundle_relations` -> `bundle_aggreations` since this is what [MSC2675](https://github.com/matrix-org/matrix-doc/pull/2675) settled on. (Also update some of the surrounded code / comments.)
* Clarifies comments for when *not* to bundle aggregations.
* Bundles aggregations for a limited `/sync` (as specified in MSC2675).
* Does *not* bundle aggregations for `/initial_sync` (as allowed in the MSC2675).
* Consistently do not include bundled aggregations for state events -- this is specified in MSC2675.

Fixes #11476

### CODE REVIEW

File: synapse/handlers/initial_sync.py [hunk 1]
```diff
@@ -165,7 +165,11 @@ async def handle_room(event: RoomsForUser) -> None:
 
                 invite_event = await self.store.get_event(event.event_id)
                 d["invite"] = await self._event_serializer.serialize_event(
-                    invite_event, time_now, as_client_event
+                    invite_event,
+                    time_now,
+                    # No need to bundle aggregations for state events.
+                    bundle_aggregations=False,
+                    as_client_event=as_client_event,
```
Comment (clokep) [inline synapse/handlers/initial_sync.py:172]: This potentially fixes an unrelated bug -- `as_client_event` was being passed in as `bundle_aggregations` instead of being passed through as a `kwarg` so `as_client_event` would always end up being `True` (the default).

File: synapse/handlers/initial_sync.py [hunk 2]
```diff
@@ -226,6 +232,8 @@ async def handle_room(event: RoomsForUser) -> None:
                 d["state"] = await self._event_serializer.serialize_events(
                     current_state.values(),
                     time_now=time_now,
+                    # No need to bundle aggregations for state events.
+                    bundle_aggregations=False,
```
Comment (clokep) [inline synapse/handlers/initial_sync.py:236]: This a bunch of places where we do this -- we do a similar check in `serialize_event` (by checking `event.is_state()`, but I think it is worthwhile to keep it here since we *know* that these are state events it is more efficient to just pass the boolean. This might be over-optimizing though and I can remove these if it is clearer!

File: synapse/events/utils.py [hunk 3]
```diff
@@ -416,26 +416,26 @@ async def serialize_event(
 
         # If MSC1849 is enabled then we need to look if there are any relations
         # we need to bundle in with the event.
```
Comment (richvdh) [inline synapse/events/utils.py:418]:
```suggestion
        # If MSC1849 is enabled then we need to look if there are any relation
        # aggregations we need to bundle in with the event.

File: synapse/rest/client/sync.py [hunk 4]
```diff
@@ -544,8 +544,13 @@ def serialize(events: Iterable[EventBase]) -> Awaitable[List[JsonDict]]:
                     event.room_id,
                 )
 
-        serialized_state = await serialize(state_events)
-        serialized_timeline = await serialize(timeline_events)
+        # No need to bundle aggregations for state events.
+        serialized_state = await serialize(state_events, bundle_aggregations=False)
+        # Only bundle aggregations if the room is limited, as clients could be
+        # missing events.
```
Comment (richvdh) [inline synapse/rest/client/sync.py:550]: This comment is confusing. Why would bundling aggregations be a problem if clients were missing events?

[The MSC doesn't really explain why we only bundle aggregations if the timeline is limited, either.]

File: synapse/handlers/initial_sync.py (hunk 1 above)
Comment (richvdh) [reply synapse/handlers/initial_sync.py:170]: ooof. maybe `bundle_relations` should be a kw-only arg to catch these problems earlier.

File: synapse/handlers/initial_sync.py (hunk 2 above)
Comment (richvdh) [reply synapse/handlers/initial_sync.py:236]: I'm ambivalent

File: changelog.d/11478.bugfix [hunk 5]
```diff
@@ -0,0 +1 @@
+Include bundled relations during a limited `/sync` request, per [MSC2675](https://github.com/matrix-org/matrix-doc/pull/2675).
```
Comment (richvdh) [inline changelog.d/11478.bugfix:1]:
```suggestion
Include bundled relation aggregations during a limited `/sync` request, per [MSC2675](https://github.com/matrix-org/matrix-doc/pull/2675).
```

File: synapse/handlers/initial_sync.py (hunk 2 above)
Comment (richvdh) [reply synapse/handlers/initial_sync.py:236]: actually I'm a bit confused. You seem to have added a bunch of these, and then removed them again in a48bd2ad. What's that about?

**Review by richvdh — COMMENTED** [review]
seems generally fine to me! A few nits and a couple of questions.

File: synapse/handlers/initial_sync.py (hunk 1 above)
Comment (clokep) [reply synapse/handlers/initial_sync.py:170]: I think that makes sense. I can make that change.

File: synapse/handlers/initial_sync.py (hunk 2 above)
Comment (clokep) [reply synapse/handlers/initial_sync.py:236]: Right, sorry should have explained that!

1. I initially added them to the places where we only serialize a state event.
2. I then added the check for `is_state()`.
3. I then removed passing the flag for places we were serializing a single event, but left it in places where we serialize many events.

As I said, maybe this is an over-optimization.

Do you think I should just leave them all in?

File: synapse/rest/client/sync.py (hunk 4 above)
Comment (clokep) [reply synapse/rest/client/sync.py:550]: I believe the idea is that if the sync is limited we include the bundled aggregations so that the clients are able to display the event (with all reactions, etc.) without having to fetch the rest of the events.

File: synapse/handlers/initial_sync.py (hunk 1 above)
Comment (clokep) [reply synapse/handlers/initial_sync.py:170]: I added this in 2f409daa8cb1834191bd5632ad3feab1584846df.

File: synapse/rest/client/sync.py (hunk 4 above)
Comment (clokep) [reply synapse/rest/client/sync.py:550]: Maybe the following is clearer:

> Only bundle aggregations if the room is limited, otherwise the client should already have all the information necessary to render events.

Does that make sense?

File: synapse/handlers/initial_sync.py (hunk 2 above)
Comment (richvdh) [reply synapse/handlers/initial_sync.py:236]: > Do you think I should just leave them all in?

I kinda do, to be honest. I think that would be clearer, and have little effect on performance.

I don't have super-strong feelings though.

Comment (richvdh) [reply synapse/handlers/initial_sync.py:236]: > I kinda do, to be honest.

wait no, the opposite. Let me try again.

I think you should let `serialize_event` sort it out, and just set `bundle_aggregations` where you need to override the behaviour. I think that would be clearer, and have little effect on performance.

File: synapse/rest/client/sync.py (hunk 4 above)
Comment (richvdh) [reply synapse/rest/client/sync.py:550]: > > Only bundle aggregations if the room is limited, otherwise the client should already have all the information necessary to render events.
>
> Does that make sense?

Right yes, that makes more sense. Alternatively:

```suggestion
        # Don't bother to bundle aggregations if the timeline is unlimited, as
        # clients will have all the necessary information.
```

File: synapse/handlers/initial_sync.py (hunk 2 above)
Comment (clokep) [reply synapse/handlers/initial_sync.py:236]: > I think you should let `serialize_event` sort it out, and just set `bundle_aggregations` where you need to override the behaviour. I think that would be clearer, and have little effect on performance.

OK. I think that's reasonable. I'll make sure that it is documented to not include them for state events too. 👍  Thanks!

Comment (clokep) [reply synapse/handlers/initial_sync.py]: Hopefully 2e76df3b023604f2b4e2d77eea895663d62ea35a makes this clear.

**Comment by clokep** [comment]
Thank you for the review @richvdh! I think I've handled all your comments! 🎉

File: synapse/events/utils.py [hunk 6]
```diff
@@ -404,7 +404,7 @@ async def serialize_event(
             event: The event being serialized.
             time_now: The current time in milliseconds
             bundle_aggregations: Whether to include the bundled aggregations for this
-                event.
+                event. Only applies to non-state events.
```
Comment (richvdh) [inline synapse/events/utils.py:407]:
```suggestion
                event. Only applies to non-state events. (State events never include bundled aggregations.)
```

**Review by richvdh — APPROVED** [review]
lgtm!

**Comment by clokep** [comment]
Thanks for the reviews! ❤️
